R/visit.R

```diff
@@ -43,3 +43,166 @@ visit_one <- function(pd_flat, funs) {
   reduce(funs, function(x, fun) fun(x),
          .init = pd_flat)
 }
+
+
+#' Propagate context to terminals
+#'
+#' Implements a very specific pre-visiting scheme, namely to propagate
+#'   indention, spaces and lag_newlines to inner token to terminals. This means
+#'   that information regarding indention, linebreaks and spaces (which is
+#'   relative in `pd_nested`) will be converted into absolute.
+#' @inherit context_towards_terminals
+#' @seealso context_towards_terminals visitors
+context_to_terminals <- function(pd_nested,
```

Comment:
I did this:

```r
context_to_terminals <- function(pd_nested) {

  if (is.null(pd_nested)) return()

  pd_transformed <- context_towards_terminals(
    pd_nested,
    pd_nested$outer_lag_newlines,
    pd_nested$outer_indent,
    pd_nested$outer_spaces
  )

  pd_transformed$child <- pmap(
    list(
      pd_transformed$child,
      pd_transformed$lag_newlines,
      pd_transformed$indent,
      pd_transformed$spaces),
    function(child, lag_newlines, indent, spaces) {
      if (is.null(child)) return(NULL)
      child[1, "outer_lag_newlines"] <- lag_newlines
      child[["outer_indent"]] <- indent
      child[nrow(child), "outer_spaces"] <- spaces
      child
    })
  pd_transformed
}
```
And changed  `context_towards_terminals` to
```r
context_towards_terminals <- function(pd_nested,
                                      outer_lag_newlines,
                                      outer_indent,
                                      outer_spaces) {
  pd_nested$indent <- pd_nested$indent + outer_indent
  pd_nested$lag_newlines <- pd_nested$lag_newlines + outer_lag_newlines
  pd_nested$spaces <-
    pd_nested$spaces + outer_spaces
  pd_nested
}
```
Which could probably done more elegantly with a `map_if()`.
In addition, we need to initialise the new columns in `create_filler`
```r
pd_flat$outer_lag_newlines <- 0
  pd_flat$outer_indent <- 0
  pd_flat$outer_spaces <- 0
```
And in all functions that create new tokens such as `add_brackets_in_pipe`.
All tests pass.
What do you think? I think initial version was more clear, did not need three new columns and had less (almost-)code duplication. Maybe  there is some more simplification I missed?